tools/metrics/tests/integration.rs

```diff
@@ -0,0 +1,268 @@
+#![cfg(feature = "nats_integration_tests")]
+
+use metrics::Args;
+
+use shared::{
+    event_msg::{event_msg::Event, EventMsg},
+    log,
+    nats_publisher_for_testing::NatsPublisherForTesting,
+    nats_server_for_testing::NatsServerForTesting,
+    nats_subjects::Subject,
+    net_conn::{self, Connection},
+    net_msg::{self, message::Msg, Metadata, Ping, Pong},
+    prost::Message,
+    rand::{self, Rng},
+    simple_logger::SimpleLogger,
+    tokio::{self, sync::watch, time::sleep},
+};
+
+use std::{
+    io::{Read, Write},
+    net::TcpStream,
+    sync::{
+        atomic::{AtomicU16, Ordering},
+        Once, OnceLock,
+    },
+    time::Duration,
+};
+
+static INIT: Once = Once::new();
+
+static NEXT_NATS_PORT: OnceLock<AtomicU16> = OnceLock::new();
+static NEXT_METRICS_PORT: OnceLock<AtomicU16> = OnceLock::new();
+
+fn setup() -> (u16, u16) {
+    INIT.call_once(|| {
+        SimpleLogger::new()
+            .with_level(log::LevelFilter::Trace)
+            .init()
+            .unwrap();
+
+        let mut rng = rand::rng();
+
+        let nats_start = rng.random_range(49152..65535); // ephemeral port range
+        let metrics_start = rng.random_range(49152..65535);
```

Comment:
don't go up to the max here
```suggestion
        let nats_start = rng.random_range(49152..65500); // ephemeral port range
        let metrics_start = rng.random_range(49152..65500);
```